### PR TITLE
Memtiles and locks

### PR DESCRIPTION
**Pull request by stephenneuendorffer**

### CODE REVIEW

**Review by fifield — APPROVED** [review]
Some of the new tests could benefit from a copyright update to 2023 AMD

**Comment by stephenneuendorffer** [comment]
> Some of the new tests could benefit from a copyright update to 2023 AMD

Fixed

File: test/dialect/AIE/badlock-vc1902.mlir [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2023 Advanced Micro Devices, Inc.
+//
+//===----------------------------------------------------------------------===//
+
+// RUN: not aiecc.py %s |& FileCheck %s
+// CHECK: error: 'AIE.lock' op lock assigned invalid id (maximum is 16)
```
Comment (keryell) [inline test/dialect/AIE/badlock-vc1902.mlir:12]:
```suggestion
// CHECK: error: 'AIE.lock' op lock assigned invalid id (maximum is 15)
```

File: lib/Dialect/AIE/IR/AIEDialect.cpp [hunk 2]
```diff
@@ -870,6 +871,14 @@ LogicalResult xilinx::AIE::LockOp::verify() {
   auto result = UsesAreAccessable::verifyTrait(*this);
   if (result.failed())
     return result;
+
+  if (getLockID().has_value()) {
+    const auto &target_model = xilinx::AIE::getTargetModel(getTileOp());
+    if (getLockID().value() >= target_model.getNumLocks())
+      return emitOpError("lock assigned invalid id (maximum is ")
+             << target_model.getNumLocks() << ")";
```
Comment (keryell) [inline lib/Dialect/AIE/IR/AIEDialect.cpp:879]:
```suggestion
             << (target_model.getNumLocks() - 1) << ")";
```

File: test/dialect/AIE/badlock-ve2802.mlir [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2023 Advanced Micro Devices, Inc.
+//
+//===----------------------------------------------------------------------===//
+
+// RUN: not aiecc.py %s |& FileCheck %s
+// CHECK: error: 'AIE.lock' op lock assigned invalid id (maximum is 64)
```
Comment (keryell) [inline test/dialect/AIE/badlock-ve2802.mlir:12]:
```suggestion
// CHECK: error: 'AIE.lock' op lock assigned invalid id (maximum is 63)
```

**Review by keryell — CHANGES_REQUESTED** [review]
Interesting!

File: lib/Dialect/AIE/IR/AIEDialect.cpp (hunk 2 above)
Comment (stephenneuendorffer) [reply lib/Dialect/AIE/IR/AIEDialect.cpp:879]: Oops.. good catch!

**Pull request merged**